File: src/scala/implementation/InnerSet_6.scala
package scala.implementation

object InnerSet_6
{
    def apply[T](element1: T, element2: T, element3: T, element4: T, element5: T, element6: T): InnerSet_6[T] = new InnerSet_6[T](element1, element2, element3, element4, element5, element6)
    def apply[T](p_innerSet: InnerSet_6[T]): InnerSet_6[T] = new InnerSet_6[T](p_innerSet)
}

class InnerSet_6[T](protected val element1: T, protected val element2: T, protected val element3: T, protected val element4: T, protected val element5: T, protected val element6: T) extends AbstractInnerSet[T]
{

    def this(p_innerSet: InnerSet_6[T]) = this(p_innerSet.element1 , p_innerSet.element2 , p_innerSet.element3 , p_innerSet.element4 , p_innerSet.element5 , p_innerSet.element6 )
    override def addElement(p_element: T): InnerSet[T] =
    {
        if (element1.equals(p_element) || element2.equals(p_element) || element3.equals(p_element) || element4.equals(p_element) || element5.equals(p_element) || element6.equals(p_element))
        {
            this
        }
        else  InnerSet_7(element1, element2, element3, element4, element5, element6, p_element)
    }

    override def addUnchecked(p_element: T): InnerSet[T] =  InnerSet_7(element1, element2, element3, element4, element5, element6, p_element)

    override def addAllElements(p_innerSet: InnerSet[T]): InnerSet[T] =  p_innerSet.addElement(element1).addElement(element2).addElement(element3).addElement(element4).addElement(element5).addElement(element6)

    override def containsElement(p_element: Object): Boolean = element1.equals(p_element) || element2.equals(p_element) || element3.equals(p_element) || element4.equals(p_element) || element5.equals(p_element) || element6.equals(p_element)

    override def containsAllElements(p_innerSet: InnerSet[T]): Boolean =  p_innerSet.containsElement(element1.asInstanceOf[Object]) && p_innerSet.containsElement(element2.asInstanceOf[Object]) && p_innerSet.containsElement(element3.asInstanceOf[Object]) && p_innerSet.containsElement(element4.asInstanceOf[Object]) && p_innerSet.containsElement(element5.asInstanceOf[Object]) && p_innerSet.containsElement(element6.asInstanceOf[Object])

    override def getElement(p_index: Int): Option[T] = p_index match{
        case 1 => Option(element1)
        case 2 => Option(element2)
        case 3 => Option(element3)
        case 4 => Option(element4)
        case 5 => Option(element5)
        case 6 => Option(element6)
        case _ => None
    }


    override def removeElement(p_element: Object): InnerSet[T] =
    {
            if(element1.equals(p_element))
         InnerSet_5(element2, element3, element4, element5, element6)
                  else             if(element2.equals(p_element))
         InnerSet_5(element1, element3, element4, element5, element6)
                  else             if(element3.equals(p_element))
         InnerSet_5(element1, element2, element4, element5, element6)
                  else             if(element4.equals(p_element))
         InnerSet_5(element1, element2, element3, element5, element6)
                  else             if(element5.equals(p_element))
         InnerSet_5(element1, element2, element3, element4, element6)
                  else             if(element6.equals(p_element))
         InnerSet_5(element1, element2, element3, element4, element5)
                        else this
    }

    override def removeAllElements(p_innerSet: InnerSet[T]): InnerSet[T] =  p_innerSet.removeElement(element1.asInstanceOf[Object]).removeElement(element2.asInstanceOf[Object]).removeElement(element3.asInstanceOf[Object]).removeElement(element4.asInstanceOf[Object]).removeElement(element5.asInstanceOf[Object]).removeElement(element6.asInstanceOf[Object])

    override def retainAllElements(p_innerSet: InnerSet[T]): InnerSet[T] =
    {
      var i: InnerSet[T] = this
            if(!p_innerSet.containsElement(element1.asInstanceOf[Object])) i = i.removeElement(element1.asInstanceOf[Object])
            if(!p_innerSet.containsElement(element2.asInstanceOf[Object])) i = i.removeElement(element2.asInstanceOf[Object])
            if(!p_innerSet.containsElement(element3.asInstanceOf[Object])) i = i.removeElement(element3.asInstanceOf[Object])
            if(!p_innerSet.containsElement(element4.asInstanceOf[Object])) i = i.removeElement(element4.asInstanceOf[Object])
            if(!p_innerSet.containsElement(element5.asInstanceOf[Object])) i = i.removeElement(element5.asInstanceOf[Object])
            if(!p_innerSet.containsElement(element6.asInstanceOf[Object])) i = i.removeElement(element6.asInstanceOf[Object])
            i
    }

    override def getSize: Int = 6

    override def iterator: InnerSetIterator[T] = InnerSetIterator(this)

    override def copy: Option[InnerSet[T]] = Option(InnerSet_6(element1: T, element2: T, element3: T, element4: T, element5: T, element6: T))

    override def clear(unused: Boolean): InnerSet[T] = InnerSet_0[T]

    override def toString : String = "{ " + element1 + ", " + element2 + ", " + element3 + ", " + element4 + ", " + element5 + ", " + element6 + " }"

    override def addAndPropagate(p_element: T, p_microSet: MicroSet[T]): InnerSet[T] =
    {
        if (element1.equals(p_element) || element2.equals(p_element) || element3.equals(p_element) || element4.equals(p_element) || element5.equals(p_element) || element6.equals(p_element))
        {
            this
        }
        else         {
            p_microSet.add(p_element)
            InnerSet_7(element1, element2, element3, element4, element5, element6, p_element)
        }    }

    override def addAllAndPropagate(p_innerSet: InnerSet[T], p_microSet: MicroSet[T]): InnerSet[T] = p_innerSet.addAllAndPropagateReverse(this, p_microSet)

    override def addAllAndPropagateReverse(p_innerSet: InnerSet[T], p_microSet: MicroSet[T]): InnerSet[T] =  p_innerSet.addAndPropagate(element1,p_microSet).addAndPropagate(element2,p_microSet).addAndPropagate(element3,p_microSet).addAndPropagate(element4,p_microSet).addAndPropagate(element5,p_microSet).addAndPropagate(element6,p_microSet)
}

File: src/cpp/implementation/save/generate/inner_set_6.cpp
#include "inner_set_6.hpp"

template<typename T, int p_to_up>
inner_set_6<T, p_to_up>::inner_set_6()
{

}

template<typename T, int p_to_up>
inner_set_6<T, p_to_up>::inner_set_6(T p_element_1, T p_element_2, T p_element_3, T p_element_4, T p_element_5, T p_element_6)
{
            a_values.element_1 = p_element_1;
            a_values.element_2 = p_element_2;
            a_values.element_3 = p_element_3;
            a_values.element_4 = p_element_4;
            a_values.element_5 = p_element_5;
            a_values.element_6 = p_element_6;
    }

template<typename T, int p_to_up>
inner_set_6<T, p_to_up>::inner_set_6(inner_set_5<T, p_to_up>* p_inner_set, T p_element)
{
            a_values.element_1 = p_inner_set->a_values.element_1;            a_values.element_2 = p_inner_set->a_values.element_2;            a_values.element_3 = p_inner_set->a_values.element_3;            a_values.element_4 = p_inner_set->a_values.element_4;            a_values.element_5 = p_inner_set->a_values.element_5;                    a_values.element_6 = p_element;
    
	delete p_inner_set;
}

template<typename T, int p_to_up>
inner_set_6<T, p_to_up>::~inner_set_6()
{

}

template<typename T, int p_to_up>
inner_set<T, p_to_up>*  inner_set_6<T, p_to_up>::copy()
{
	return this;
}

template<typename T, int p_to_up>
inner_set<T, p_to_up>* inner_set_6<T, p_to_up>::add_element(T p_element)
{
	if(!contains_element(p_element))
		if(this->a_to_up == 6)
			return new inner_array_set<T, p_to_up>(this, p_element);
		else
			return new inner_set_7<T, p_to_up>(this, p_element);
	else
		return this;
}

template<typename T, int p_to_up>
inner_set<T, p_to_up>* inner_set_6<T, p_to_up>::add_all_elements(inner_set<T, p_to_up>* p_inner_set)
{
  return p_inner_set->add_element(a_values.element_1)->add_element(a_values.element_2)->add_element(a_values.element_3)->add_element(a_values.element_4)->add_element(a_values.element_5)->add_element(a_values.element_6);
}

template<typename T, int p_to_up>
bool inner_set_6<T, p_to_up>::contains_element(T p_element)
{
    return  a_values.element_1 == p_element ||  a_values.element_2 == p_element ||  a_values.element_3 == p_element ||  a_values.element_4 == p_element ||  a_values.element_5 == p_element ||  a_values.element_6 == p_element;
}

template<typename T, int p_to_up>
bool inner_set_6<T, p_to_up>::contains_all_elements(inner_set<T, p_to_up>* p_inner_set)
{
   return  p_inner_set->contains_element(a_values.element_1)  &&   p_inner_set->contains_element(a_values.element_2)  &&   p_inner_set->contains_element(a_values.element_3)  &&   p_inner_set->contains_element(a_values.element_4)  &&   p_inner_set->contains_element(a_values.element_5)  &&   p_inner_set->contains_element(a_values.element_6) ;
}

template<typename T, int p_to_up>
T inner_set_6<T, p_to_up>::get_element(int p_index)
{
    return reinterpret_cast<T*>(&a_values)[p_index];
}

template<typename T, int p_to_up>
inner_set<T, p_to_up>* inner_set_6<T, p_to_up>::remove_element(T p_element)
{
                if(a_values.element_1 == p_element)
                {
        	delete this;
        	return new inner_set_5<T, p_to_up>(a_values.element_2a_values.element_3a_values.element_4a_values.element_5a_values.element_6);
    	}
         else                 if(a_values.element_2 == p_element)
                {
        	delete this;
        	return new inner_set_5<T, p_to_up>(a_values.element_1a_values.element_3a_values.element_4a_values.element_5a_values.element_6);
    	}
         else                 if(a_values.element_3 == p_element)
                {
        	delete this;
        	return new inner_set_5<T, p_to_up>(a_values.element_1a_values.element_2a_values.element_4a_values.element_5a_values.element_6);
    	}
         else                 if(a_values.element_4 == p_element)
                {
        	delete this;
        	return new inner_set_5<T, p_to_up>(a_values.element_1a_values.element_2a_values.element_3a_values.element_5a_values.element_6);
    	}
         else                 if(a_values.element_5 == p_element)
                {
        	delete this;
        	return new inner_set_5<T, p_to_up>(a_values.element_1a_values.element_2a_values.element_3a_values.element_4a_values.element_6);
    	}
         else                 if(a_values.element_6 == p_element)
                {
        	delete this;
        	return new inner_set_5<T, p_to_up>(a_values.element_1a_values.element_2a_values.element_3a_values.element_4a_values.element_5);
    	}
                        else
            return this;
}

template<typename T, int p_to_up>
inner_set<T, p_to_up>* inner_set_6<T, p_to_up>::remove_all_elements(inner_set<T, p_to_up>* p_inner_set)
{
  return p_inner_set->remove_element(a_values.element_1)->remove_element(a_values.element_2)->remove_element(a_values.element_3)->remove_element(a_values.element_4)->remove_element(a_values.element_5)->remove_element(a_values.element_6);
}

template<typename T, int p_to_up>
inner_set<T, p_to_up>* inner_set_6<T, p_to_up>::retain_all_elements(inner_set<T, p_to_up>* p_inner_set)
{
  inner_set<T, p_to_up>* r_inner_set = inner_set_0<T, p_to_up>::EMPTY;

		if(p_inner_set->contains_element(a_values.element_1))
		 r_inner_set = r_inner_set->add_element(a_values.element_1);
	
	 	if(p_inner_set->contains_element(a_values.element_2))
		 r_inner_set = r_inner_set->add_element(a_values.element_2);
	
	 	if(p_inner_set->contains_element(a_values.element_3))
		 r_inner_set = r_inner_set->add_element(a_values.element_3);
	
	 	if(p_inner_set->contains_element(a_values.element_4))
		 r_inner_set = r_inner_set->add_element(a_values.element_4);
	
	 	if(p_inner_set->contains_element(a_values.element_5))
		 r_inner_set = r_inner_set->add_element(a_values.element_5);
	
	 	if(p_inner_set->contains_element(a_values.element_6))
		 r_inner_set = r_inner_set->add_element(a_values.element_6);
	
	 
    if(this->get_size() != 0)
      delete this;

  return r_inner_set;
}


template<typename T, int p_to_up>
inner_set<T, p_to_up>* inner_set_6<T, p_to_up>::add_and_propagate(T p_element, micro_set<T, p_to_up>* p_micro_set)
{
	if( a_values.element_1 == p_element  ||   a_values.element_2 == p_element  ||   a_values.element_3 == p_element  ||   a_values.element_4 == p_element  ||   a_values.element_5 == p_element  ||   a_values.element_6 == p_element )
		return this;
	else
  	{
  		p_micro_set->insert(p_element);

		return new inner_set_7<T, p_to_up>(this, p_element);
	}
}

template<typename T, int p_to_up>
inner_set<T, p_to_up>* inner_set_6<T, p_to_up>::add_all_and_propagate(inner_set<T, p_to_up>* p_inner_set, micro_set<T, p_to_up>* p_micro_set)
{
  return p_inner_set->add_all_and_propagate_reverse(this, p_micro_set);
}

template<typename T, int p_to_up>
inner_set<T, p_to_up>* inner_set_6<T, p_to_up>::add_all_and_propagate_reverse(inner_set<T, p_to_up>* p_inner_set, micro_set<T, p_to_up>* p_micro_set)
{
	return p_inner_set->add_and_propagate(a_values.element_1, p_micro_set)->add_and_propagate(a_values.element_2, p_micro_set)->add_and_propagate(a_values.element_3, p_micro_set)->add_and_propagate(a_values.element_4, p_micro_set)->add_and_propagate(a_values.element_5, p_micro_set)->add_and_propagate(a_values.element_6, p_micro_set);
}

template<typename T, int p_to_up>
int inner_set_2<T, p_to_up>::get_size()
{
    return 6;
}
